-

SergeyRogonov opened the pull request:
The first iteration with its all functionality
add parser.py, news_articles.py, .gitignore
changed rss_reader.py, requirements.txt

SergeyRogonov committed:
The first iteration with its all functionality
add parser.py, news_articles.py, .gitignore
changed rss_reader.py, requirements.txt

HenadziStantchik commented:
Please fix this PR's name so it corresponds the one in README.md

HenadziStantchik commented on final_task/rss_reader/requirements.txt:
```diff
@@ -0,0 +1,14 @@
+pylint==2.4.3
+six==1.13.0
+soupsieve==1.9.5
+wrapt==1.11.2
```
There are many libs that you do not use here. It is better to clean this up.

SergeyRogonov replied:
fixed

HenadziStantchik commented:
Here is some feedback on your app work:

1. when using `--json` option, your app has encoding problem:

![rogonov](https://user-images.githubusercontent.com/43743521/68856025-94684300-06f0-11ea-8e2c-4a7cfff6e4e7.png)

2. If `--limit` is larger than feed size, `StopIteration` is printed with traceback to stdout.

HenadziStantchik commented on final_task/rss_reader/news_articles.py:
```diff
@@ -0,0 +1,29 @@
+
+class NewsArticle:
+    """Contain data about an article"""
+    def __init__(self, news_outlett_name, news_title, pub_date, news_link,
```
Consider using `dataclass` here

SergeyRogonov replied:
fixed. it really looks better

HenadziStantchik commented on final_task/rss_reader/rss_reader.py:
```diff
@@ -0,0 +1,72 @@
+    parser.add_argument("--json", help="Print result as JSON in stdout", action="store_true")
+    parser.add_argument("--verbose", help="Outputs verbose status messages", action="store_true")
+    parser.add_argument("--limit", help="Limit news topics if this parameter provided")
+    args = parser.parse_args()
```
Better move this to a separate function

SergeyRogonov replied:
fixed

HenadziStantchik commented on final_task/rss_reader/rss_reader.py:
```diff
@@ -0,0 +1,72 @@
+    return response
+
+
+def output_format(news_articles, json_output=False):
```
Please be consistent with typehints. It is better to either typehint everywhere(preferrable) or nowhere(still ok)

SergeyRogonov replied:
removed it for now

SergeyRogonov committed:
2nd iteration is complete

added Usage, Installation and JSON stracture to README.md
changed setup.py for 2nd iteration
added dist/rss_reader-0.2.0.tar.gz for second iteration
now using dataclass in news_articles
fixed problem when --limit is greater than number of news articles
temporarily removed type hints
added function get_args to rss_reader.py

HenadziStantchik commented on final_task/README.md:
````diff
@@ -1,3 +1,42 @@
+    "img_alt": "image alternative if exist",
+    "img_src": "image link if exist",
+}
+```
````
Please add information how to launch your app using setuptools

SergeyRogonov replied:
I actually have a problem with it. To launch from source you can simply type "rss_reader.py [-h] [--version] [--json] [--verbose] [--limit LIMIT] source". But for installation via setuptools I can't name entry_point console_script from setup.py the same name as my entry point rss_reader.py. With other names it works fine. For now it works like "rss_reader [-h] [--version] [--json] [--verbose] [--limit LIMIT] source". I suppose it isn't OK so can you give me any tips, please. Or should I leave it be and just add the information about how to launch the app using setuptools?

HenadziStantchik replied:
Try to bypass this problem using `sys.path.append` if you have troubles when import does not work either for setuptools or launching from source

HenadziStantchik commented on final_task/rss_reader/requirements.txt:
```diff
@@ -1,14 +1,4 @@
-astroid==2.3.3
 beautifulsoup4==4.8.1
 bs4==0.0.1
```
A question:

What is the difference between bs4 and beautifulsoup4? Do you really need both of them?

SergeyRogonov replied:
My bad. It seems when you install like "pip install bs4" it installs beautifulsoup4 and this dummy package. I will remove it.

SergeyRogonov committed:
iteration 3

HenadziStantchik commented on final_task/rss_reader/news_articles.py:
```diff
@@ -24,5 +24,16 @@ def __repr__(self):
+    def __eq__(self, other):
+        if isinstance(other, NewsArticle):
+            return self.__key() == other.__key()
+        return NotImplemented
```
A question: Why do you need to return this here?

SergeyRogonov replied:
Initially I find it on stackoverflow as a good practice to implement __hash__(). In docs I found out that "If class does not define an __eq__() method it should not define a __hash__() operation either". So I decided to implement __hash__() with __eq__() together. And about "return NotImplemented" I found out that if the other object, we are comparing with, is not an instance of NewsArticle class, then this method goes back to built-in behavior.
Actually, for my project it is not necessary, so is it better I remove it? 
p.s. Do I get it all right?

HenadziStantchik commented on final_task/README.md:
```diff
@@ -17,14 +17,21 @@ optional arguments:
   --json         Print result as JSON in stdout
   --verbose      Outputs verbose status messages
   --limit LIMIT  Limit news topics if this parameter provided
+  --date         Read cached news for provided URL. If "ALL" provided - prints all cached news for this date
```
Please add a bit more info about how do you store cached news. What do you use and etc.

SergeyRogonov replied:
Sure. Will fix it.

HenadziStantchik commented:
I did not discover any issues with Iteration 3 work.

SergeyRogonov committed:
iteration 4 and tests\
renamed parser.py to rss_parser.py because of name conflict
add tests: test_cache.py, test_converter.py, test_rss_reader.py, test_parser.py in test dir
add converter.py that can convert to html and pdf. it uses jinja2 nd weasyprint
add template.html for convertion to other formats
add argument checking for --date and --limit in rss_parser.py
split update_cache() from cache.py -> save_cache() and update_cache()
fixed problem when cache storage. now it is stored in /tmp/rss_reader/cache.
previously it was in the directory where the reader was running

HenadziStantchik commented on final_task/rss_reader/requirements.txt:
```diff
@@ -0,0 +1,21 @@
+soupsieve==1.9.5
+tinycss2==1.0.2
+WeasyPrint==50
+webencodings==0.5.1
```
There is too much redundant libs here. Usually only libs that are needed for your application _directly_ specified in this file.

SergeyRogonov replied:
Oh, I see. I should not list all of them here even if most of them were installed by libs that I use directly?

HenadziStantchik replied:
>Oh, I see. I should not list all of them here even if most of them were installed by libs that I use directly?

Yes. The same for `dependencies` in `setup.py`

SergeyRogonov replied:
I get it. Thanks

HenadziStantchik commented on final_task/rss_reader/cache.py:
```diff
@@ -0,0 +1,105 @@
+            if cache.get(url):
+                for article in cache[url]:
+                    if len(news_articles) == limit:
+                        break
```
It is better to use slices (`[:]`) for this instead of comparing articles length to the limit on each iteration.

SergeyRogonov replied:
Thanks. I fixed it.

HenadziStantchik commented:
Here is feedback on your application:
Iteration 1:
Does not work using `python rss_reader.py ....`
![image](https://user-images.githubusercontent.com/43743521/70054141-fdecba80-15e7-11ea-94b4-9a60b3d0ae47.png)

Iteration 2:
Installs, but does not work:
![image](https://user-images.githubusercontent.com/43743521/70053516-abf76500-15e6-11ea-8d8c-42571e585378.png)

Iteration 3:
Could not be tested.

Iteration 4:
Could not be tested.

Tests are decently written, coverage is high.
The code is decently structured, readable, and absolutely correspond to the PEP8 guidelines.
Commit messages are informative and correspond to the recommended commit message style(good job)
Using templates for html is a plus.
Unfortunately application is not in working condition. It was tested both on a clean machine and inside isolated docker container

SergeyRogonov committed:
fixed problems with imports which caused program stopped working
renamed test and static directories to tests and templates
